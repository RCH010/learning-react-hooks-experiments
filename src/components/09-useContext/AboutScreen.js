import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const AboutScreen = () => {

    const {user, setUser} = useContext(UserContext)
    
    const handleClick = () =>{
        setUser({})
    }

    return (
        <div>
            <h1>About Screen</h1> 
            <hr/>
            <pre>{JSON.stringify(user, '-', 3)}</pre>

            <button
                className='btn btn-warning'
                onClick={handleClick}
            >
               Logout
            </button>
        </div>
    )
}
