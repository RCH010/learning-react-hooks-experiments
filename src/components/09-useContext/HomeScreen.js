import React, { useContext } from 'react'
import { UserContext } from './UserContext';

export const HomeScreen = () => {

    const {user, setUser} = useContext(UserContext);
    console.log(user)

    return (
        <div>
            <h1>Home Screen</h1> 
           <hr/>
           <pre>
               {JSON.stringify(user, null, 3)}
           </pre>
        </div>
    )
}
