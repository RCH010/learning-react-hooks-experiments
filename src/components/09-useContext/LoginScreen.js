import React, { useContext } from 'react'
import { useForm } from '../../hooks/useForm'
import { UserContext } from './UserContext'

const getNewID = () => {
    return new Date().getTime();
}

export const LoginScreen = () => {
    // el UserContext es el nombre que le definimos, hay que importarlo
    //  el use context, busca en el arbol de componentes el que estoy pidiendo (UserContext)
    const {user, setUser} = useContext(UserContext)
    const [{name, email}, handleInputChange, reset] = useForm({
        name:'',
        email:''
    })
    
    const handleSubmit = (e) => {
        e.preventDefault();
        if(name.trim().length <= 1 || email.trim().length <= 1) { return; }
        setUser({
            id: getNewID(),
            name: name,
            email: email
        })
        reset();
    }

    return (
        <div className='row justify-content-center'>
            <div className='col-7'>
                <div className='card text-white bg-dark text-center m-5'>
                    <div className='card-body'>
                        <h1 className='h2 card-title'>Iniciar Sesión</h1>
                        <div className='card-text my-2'>
                            <form onSubmit={handleSubmit}>
                                <div className='form-group'>
                                    <input 
                                        className="form-control" 
                                        type='text' 
                                        placeholder='Nombre'
                                        name='name'
                                        value={name}
                                        onChange={handleInputChange}
                                    />
                                </div>
                                <div className='form-group'>
                                    <input 
                                        className="form-control"
                                        type='email'
                                        placeholder='Correo Electrónico'
                                        name='email'
                                        value={email}
                                        onChange={handleInputChange}
                                    />
                                </div>
                                <button type="submit" className="btn btn-primary">Inicar Sesión</button>
                            </form>
                        </div>
                    </div>
                </div>
                <pre>{JSON.stringify(user)}</pre>
            </div>
        </div>
    )
}
