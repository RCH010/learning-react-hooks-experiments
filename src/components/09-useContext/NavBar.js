import React from 'react'
// import { Link, NavLink } from 'react-router-dom'
import { NavLink } from 'react-router-dom'

export const NavBar = () => {
    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
            
            <button className="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarText" aria-controls="navbarText" aria-expanded="false" aria-label="Toggle navigation">
            <span className="navbar-toggler-icon"></span>
            </button>

            <div className="collapse navbar-collapse" id="navbarText">
            <ul className="navbar-nav mr-auto">
                <li className="nav-item">
                    <NavLink exact activeClassName='active' className='nav-link' to='./'>Home</NavLink>
                </li>
                <li className="nav-item">
                    <NavLink exact activeClassName='active' className='nav-link' to='./about'>About</NavLink>
                </li>
                <li className="nav-item">
                    <NavLink exact activeClassName='active' className='nav-link' to='./login'>Login</NavLink>
                </li>
            </ul>
            
            </div>
        </nav>
    )
}
// el <Link> y el <NavLink> son lo mismo, solo que con navlink podemos especificar la clase que va si esta activa
