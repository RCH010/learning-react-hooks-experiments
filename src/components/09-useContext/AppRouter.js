import React from 'react'
import {
    BrowserRouter as Router,
    Switch,
    Route,
    Redirect,
  } from "react-router-dom";
import { AboutScreen } from './AboutScreen';
import { HomeScreen } from './HomeScreen';
import { LoginScreen } from './LoginScreen';
import { NavBar } from './NavBar';

export const AppRouter = () => {
    return (
        <Router>
            <div>
                <NavBar />
                <div className='container'>
                    <Switch>
                        {/* con el prop exact, hace que sea literal */}
                        <Route exact path='/about' component={ AboutScreen }/>
                        <Route exact path='/login' component={ LoginScreen }/>
                        <Route exact path='/' component={ HomeScreen }/>
                        <Redirect to='/'/>  {/* Cuelquier otro cae aqui */}
                    </Switch>
                </div>
            </div>
        </Router>
    )
}
