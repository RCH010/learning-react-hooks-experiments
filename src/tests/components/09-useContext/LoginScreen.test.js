import { mount } from 'enzyme';
import React from 'react';
import { LoginScreen } from '../../../components/09-useContext/LoginScreen';
import { UserContext } from '../../../components/09-useContext/UserContext';

describe('Tests in <LoginScreen />', () => {
    // const user = {
    //     id: 0,
    //     name: '',
    //     email: ''
    // }
    const setUser = jest.fn();
    const wrapper = mount(
        <UserContext.Provider value={{setUser}}>
            <LoginScreen/>
        </UserContext.Provider>
    );
    test('should match with snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should not setUser if form is empty', () => {
        let inputs = wrapper.find('input');
        inputs.forEach(input => {
            input.simulate('change', {target: {name: input.prop('name'), value:'h'}})
        })
        wrapper.find('button').simulate('click');
        inputs = wrapper.find('input');
        inputs.forEach(input => {
            expect(input.prop('value')).toBe('h');
        })
    });

    // test('should change user values when form is submited', () => {
    //     const inputs = wrapper.find('input');
    //     inputs.forEach(input => {
    //         input.simulate('change', {target: {name: input.prop('name'), value:'hello'}})
    //     })
    //     const formsubmit = wrapper.find('form').prop('onSubmit');
    //     formsubmit({preventDefault(){}});
    //     expect(wrapper.find('pre').text()).toBe('');

    // })

})
