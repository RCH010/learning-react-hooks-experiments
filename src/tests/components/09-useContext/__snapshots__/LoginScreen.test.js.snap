// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <LoginScreen /> should match with snapshot 1`] = `
<div
  className="row justify-content-center"
>
  <div
    className="col-7"
  >
    <div
      className="card text-white bg-dark text-center m-5"
    >
      <div
        className="card-body"
      >
        <h1
          className="h2 card-title"
        >
          Iniciar Sesión
        </h1>
        <div
          className="card-text my-2"
        >
          <form
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <input
                className="form-control"
                name="name"
                onChange={[Function]}
                placeholder="Nombre"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <input
                className="form-control"
                name="email"
                onChange={[Function]}
                placeholder="Correo Electrónico"
                type="email"
                value=""
              />
            </div>
            <button
              className="btn btn-primary"
              type="submit"
            >
              Inicar Sesión
            </button>
          </form>
        </div>
      </div>
    </div>
    <pre />
  </div>
</div>
`;
